contracts/tunnel/BaseRootTunnel.sol

```diff
@@ -0,0 +1,203 @@
+import {IStateSender} from "../root/StateSender/IStateSender.sol";
+import {RLPReader} from "../lib/RLPReader.sol";
+import {MerklePatriciaProof} from "../lib/MerklePatriciaProof.sol";
+import {ICheckpointManager} from "../root/ICheckpointManager.sol";
+import {RLPReader} from "../lib/RLPReader.sol";
+import {MerklePatriciaProof} from "../lib/MerklePatriciaProof.sol";
+import {Merkle} from "../lib/Merkle.sol";
+
+abstract contract BaseRootTunnel is AccessControlMixin {
+    using RLPReader for bytes;
+    using RLPReader for RLPReader.RLPItem;
+    using Merkle for bytes32;
+    using SafeMath for uint256;
+
+    bytes32 public constant SENDER_MANAGER_ROLE = keccak256("SENDER_MANAGER_ROLE");
+    bytes32 public constant RECEIVER_MANAGER_ROLE = keccak256("RECEIVER_MANAGER_ROLE");
+  
+    // keccak256(SendMessage(bytes))
+    bytes32 public constant SEND_MESSAGE_EVENT_SIG = 0x69331a26390a78389351d9264561047b757948e9b0a1a9dfc6bec2b230b50953;
+
+    // state sender contract
+    IStateSender public stateSender;
+    // root chain manager
+    ICheckpointManager public checkpointManager;
+    // child tunnel contract which receives and sends messages 
+    address public childTunnel;
+    // storage to avoid duplicate exits
+    mapping(bytes32 => bool) public processedExits;
+
+    constructor() public {
+      _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /**
+     * @notice Set the state sender, callable only by admins
+     * @dev This should be the state sender from plasma contracts
+     * It is used to send bytes from root to child chain
+     * @param newStateSender address of state sender contract
+     */
+    function setStateSender(address newStateSender)
+        external
+        only(DEFAULT_ADMIN_ROLE)
+    {
+        stateSender = IStateSender(newStateSender);
+    }
+
+    /**
+     * @notice Set the checkpoint manager, callable only by admins
+     * @dev This should be the plasma contract responsible for keeping track of checkpoints
+     * @param newCheckpointManager address of checkpoint manager contract
+     */
+    function setCheckpointManager(address newCheckpointManager)
+        external
+        only(DEFAULT_ADMIN_ROLE)
+    {
+        checkpointManager = ICheckpointManager(newCheckpointManager);
+    }
+
+    /**
+     * @notice Set the child chain tunnel, callable only by admins
+     * @dev This should be the contract responsible to receive data bytes on child chain
+     * @param newChildTunnel address of child tunnel contract
+     */
+    function setChildTunnel(address newChildTunnel)
+        external
+        only(DEFAULT_ADMIN_ROLE)
+    {
+        require(newChildTunnel != address(0x0), "MessageTunnel: INVALID_CHILD_TUNNEL_ADDRESS");
+        childTunnel = newChildTunnel;
+    }
+  
+    // send message from L1 to L2
+    function sendMessage(bytes calldata data) external only(SENDER_MANAGER_ROLE) {
+        stateSender.syncState(childTunnel, data);
+    }
```

Comment:
Same comment as above, lets make this internal and have implementation contract determine when it needs to be called